[Junha Park] ip

B1LLP4RK opened the pull request:
# Duke, your personal secretary
> "Give me six hours to chop down a tree and I will spend the first four sharpening the axe." - Abraham Lincoln

Knowing exactly what you have to do will boost your productivity:rocket:!
**Duke** is a CLI based todo-List app that helps keep you track of what to do at which time, until when.
You can download the app [here](https://github.com/B1LLP4RK/ip).

## commands
1. `todo {task name}`
   - adds a task to the tasklist
2. `event {event name} \at {when}`
   - adds a task that starts at a time to the tasklist
3. `deadline {event name} \by {when}`
   - adds a task that is due at a particular time to the tasklist
4. `mark {task number}`/`unmark {task number}`
   - marks/unmarks a task done
5. `delete {task number}`
   - delete a particular task
6. `find {keyword}`
   - shows the task containing the keyword, case insensitive
## development roadmap
- [x] enable search using a **keyword**
- [ ] enable changing **task orders**
- [ ] enable setting **priority** to tasks
- [ ] enable **priority-based ordering**
## known issues
- giving multiple keywords delimited by space for `find` crashes the app (will be fixed soon)
## patch notes
- no patch yet
## to fellow CS students
Feel free to fork and make pull requests to this project! The code is clean enough for a novice to understand what's going on.:smiley: The main logic is roughly:
```java
while (!over) {
            String userInput = myScanner.nextLine();
            Command command = Parser.parse(userInput, taskList, over);
            command.execute();
}           
```

Commit:
Add Gradle support

B1LLP4RK committed:
greet, echo, exit

B1LLP4RK committed:
enable list function

B1LLP4RK committed:
Add the ability to mark tasks as done.

B1LLP4RK committed:
add ToDos, Events, Deadlines

B1LLP4RK committed:
handle errors

B1LLP4RK committed:
fix a bug where wrong response was given to 'blah'

B1LLP4RK committed:
impletment task deletion

B1LLP4RK committed:
automate testing

B1LLP4RK committed:
add save functionality

B1LLP4RK committed:
store date as DateTime class

B1LLP4RK committed:
Merge branch 'branch-Level-7'

B1LLP4RK committed:
Merge branch 'branch-Level-8'

B1LLP4RK committed:
fix the runtime error when entering dates

B1LLP4RK committed:
refactor code into OOP style

B1LLP4RK committed:
organize code into packages

B1LLP4RK committed:
Add Junit tests

B1LLP4RK committed:
A-Jar

B1LLP4RK committed:
build jar file

B1LLP4RK committed:
add javadoc

B1LLP4RK committed:
fix coding standard

B1LLP4RK committed:
add find using keyword

B1LLP4RK committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Over.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Response.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Todo.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/CommandBye.java

B1LLP4RK committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/command/CommandUnclear.java

B1LLP4RK committed:
add find

NicsunXnus commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /**
+     * returns a string representation of the deadline to be seen by users.
```
Just a minor comment, I think JavaDoc comments should start off with a Capital letter.
From the[ SE-guide](https://se-education.org/guides/conventions/java/intermediate.html#comments):
```
/**
 * Returns lateral location of the specified position.
 * If the position is unset, NaN is returned.
 *
 * @param x  X coordinate of position.
 * @param y Y coordinate of position.
 * @param zone Zone of position.
 * @return Lateral location.
 * @throws IllegalArgumentException  If zone is <= 0.
 */
public double computeLocation(double x, double y, int zone)
    throws IllegalArgumentException {
  ...
}
```

NicsunXnus commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return a string representation of the deadline to be used for data storage.
+     */
+    @Override
+    public String toData() {
+        String isFinishedData;
+        if (super.finished) {
+            isFinishedData = "1";
+        } else {
+            isFinishedData = "0";
+        }
+        return "D:" + isFinishedData + ":" + super.content + ":" + date.format(DateTimeFormatter.ISO_DATE);
+    }
```
Some comments about this:
1) I think isFinishedData could be renamed to something else since it sounds like a boolean variable. Maybe try "doneStatus" or "finishedStatus"?
2) Since you are using strings to show boolean states maybe you can change the code to look something like this. I feel that simply assigning the isFinishedData to a number does not tell readers much immediately.:
```
@Override
    public String toData() {
        String finishedStatus;
        String completedStatus = "1";
        String incompleteStatus = "0";
        if (super.finished) {
            doneStatus = completedStatus;
        } else {
           doneStatus = incompleteStatus;
        }
        return "D:" + finishedStatus + ":" + super.content + ":" + date.format(DateTimeFormatter.ISO_DATE);
    }
```

B1LLP4RK replied:
I am actually honored to receive this suggestion. This is so good. @NicsunXnus

NicsunXnus commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+        storage.readData(taskList);
+        Scanner myScanner = new Scanner(System.in);
+        System.out.println("Hello from\n" + Response.LOGO);
+        Ui.wrapPrint(Response.RESPONSE_WELCOME);
+        Over over = new Over();
```
Should there be some comments to explain what these two statements does?

NicsunXnus commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+                Command command = Parser.parse(userInput, taskList, over);
+                command.execute();
+            } catch (NumberFormatException e) {
+                //mark command
+                //unmark command
```
I think you may have left something out here

NicsunXnus commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+                Ui.wrapPrint(Response.RESPONSE_MISSINGDATE);
+                //event command
+                Ui.wrapPrint(Response.RESPONSE_MISSINGTIME);
+                //} catch (DukeCannotUnderstandException e) {
```
I think you forgot to remove this comment.

NicsunXnus commented on src/main/java/duke/Over.java:
```diff
@@ -0,0 +1,22 @@
+ * Encapsulates boolean value to be used like a mutable object.
+ */
+public class Over {
+    boolean isOver;
```
Should this be made public? 
From [SE-Guide](url): Class variables should never be declared public unless the class is a data class with no behavior. This rule does not apply to constants.

NicsunXnus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+package duke;
+
+import duke.command.*;
```
Should this be a wildcard import?
From [SE-Guide](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements):  Imported classes should always be listed explicitly.

NicsunXnus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+     */
+    public static Command parse(String userInput, TaskList taskList, Over over) throws NoTimeGivenException {
+        String[] words = userInput.split(" ");
+        String firstWord = words[0];
+        if (firstWord.equals("bye")) {
```
I think a line break between this two lines will improve readability

NicsunXnus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+        String[] words = userInput.split(" ");
+        String firstWord = words[0];
+        if (firstWord.equals("bye")) {
+            return new CommandBye(over);
+        } else if (firstWord.equals("list")) {
+            return new CommandList(taskList);
+        } else if (firstWord.equals("mark") && words.length == 2) {
+            int taskNo = Integer.parseInt(words[1]);
+            return new CommandMark(taskList, taskNo);
+        } else if (firstWord.equals("unmark") && words.length == 2) {
+            int taskNo = Integer.parseInt(words[1]);
+            return new CommandUnmark(taskList, taskNo);
+        } else if (firstWord.equals("todo")) {
+            String todoContent = words[1];
+            for (int i = 2; i < words.length; i++) {
+                todoContent = todoContent + " " + words[i];
+            }
```
Maybe you could try a ```switch-case block``` instead, to make the code seem neater?
For example:
```
switch (firstWord) {
case "bye":
    //
case "list":
    //
default:
    //
    break;
}
```

NicsunXnus commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+            }
+            return new CommandTodo(taskList, todoContent);
+        } else if (firstWord.equals("deadline")) {
+            String deadlineContent = words[1];
+            for (int i = 2; i < words.length; i++) {
+                if (words[i].equals("/by")) {
+                    break;
+                }
+                deadlineContent = deadlineContent + " " + words[i];
```
Wouldn't it be better if you could group all of this logic into one method? It would make the code more succinct and easier to read. You could also do the same thing for the rest of the code.
```suggestion
            //String deadlineContent = words[1];
            //for (int i = 2; i < words.length; i++) {
            //    if (words[i].equals("/by")) {
            //        break;
            //   }
            //    deadlineContent = deadlineContent + " " + words[i];
                deadlineContent = makeDeadlineContent(words);
```

B1LLP4RK replied:
Yes!

NicsunXnus reviewed (commented):
Hi, I made a couple of reviews and I hope that they don't seem too harsh but I felt that they are needed to made the code better. By the way, I like your use of Enums for some parts of your code.

B1LLP4RK committed:
Merge branch 'add-gradle-support' of https://github.com/B1LLP4RK/ip into branch-A-Gradle

B1LLP4RK committed:
setup gradle

B1LLP4RK committed:
setup checkstyle

zechajw commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * returns a sting representation of the event to be used for data storage.
```
I think there is a typo here and it should be "string" instead of "sting"

zechajw commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    @Override
+    public String toData() {
+        String isFinishedData;
```
Should the String name here be "isFinishedData"? According to the coding convention, a variable name starting with "is" is commonly used for boolean values rather than String values.

zechajw commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,62 @@
+ * Provides abstraction of an task in the todoList
+ */
+public abstract class Task {
+    protected boolean finished;
```
Perhaps you could change the variable name here to isFinished to align with the coding convention for boolean values?

zechajw commented on src/main/java/duke/command/CommandBye.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    /**
+     * Execution of this command.
```
Perhaps a clearer description of the method here could be better? The description could provide more additional information other than what is already available in the method name.

zechajw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,88 @@
+            return new CommandDelete(taskList, taskNo);
+        } else if (firstWord.equals("find") && words.length == 2) {
+            String wordSecond = words[1];
+            System.out.println("reached line 58");
```
Perhaps these lines could be deleted in production code so that it doesn't appear for the user?

zechajw reviewed (commented):
Looks really well done so far! 👍 

Overall, I found that the code was generally easy to read except for a few lapses in coding convention that I pointed out in my comments! Otherwise, good work so far!

B1LLP4RK committed:
setup GUI

juayhee commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Creates an Event
+     *
+     * @param content description of the event
```
I think the formatting could be improved if the Javadocs started with capital letters in sentences.

juayhee commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,108 @@
+ * Encapsulates a list of a {@link Task}
+ */
+public class TaskList {
+    private final ArrayList<Task> todo;
```
Perhaps the variable name may be changed to more accurately represent that this is a collection of Tasks. May I suggest todoList?

juayhee commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,13 @@
+hello
+hello again
+todo ride a horse
+deadline give me a ride /by Sunday
+Event attend the lecture /at Sun 2~4pm
+list
+mark 1
+mark 2
+mark 3
+unmark 2
+list
+bye
+
```
Comprehensive tests!

juayhee reviewed (commented):
I found your code easy to read and understand. I feel like minimal changes were needed. Looks great!

B1LLP4RK committed:
Use assert feature

Current code does not utilise any assert feature at all.

Using assert can help detect whether the underlying assumptions of the
program states hold true. Therefore, bugs are found before they turn out
 to be more complicated issue.

B1LLP4RK committed:
Refactor Codes

The codebase lacks documentation and has a lot of nested loops and
repeated parts.

Documentation helps people using the package have a clear understanding
of the methods and classes and use them correctly. It reduces future
development time. Replacing duplicate code makes the code more succinct
and easier to change implementation and debug in the future.

B1LLP4RK committed:
Merge pull request #1 from B1LLP4RK/branch-A-CodeQuality

Refactor Codes

B1LLP4RK committed:
Merge branch 'master' into branch-A-Assertions

B1LLP4RK committed:
Merge pull request #2 from B1LLP4RK/branch-A-Assertions

Use assert feature

B1LLP4RK committed:
Add tagging feature

B1LLP4RK committed:
Set theme jekyll-theme-slate

B1LLP4RK committed:
update Readme

B1LLP4RK committed:
Merge branch 'branch-C-Tagging'

B1LLP4RK committed:
Merge remote-tracking branch 'origin/master'

B1LLP4RK committed:
update Readme

B1LLP4RK committed:
fix the bug that event and deadline commands are not working

B1LLP4RK committed:
Improve User Guide

B1LLP4RK committed:
fix the issue that User Guide is not correctly rendered in the web page

B1LLP4RK committed:
Fix issue with the naming convention of test methods

B1LLP4RK committed:
Informs user to enter valid number when user enters invalid person index

B1LLP4RK committed:
Fix incorrect parsing of Deadline and Event command string

B1LLP4RK committed:
Fix truncation of todo list when the there are many tasks

B1LLP4RK committed:
Inform 'argument' missing when todo command is used without any content

B1LLP4RK committed:
Correct the dividing keyword '\by' to '/by' in README.md
